fix!: Fix eth tx hashes in json-rpc responses

yihuang opened the pull request:
## Description

Fix eth tx hashes in json-rpc responses

Closes: #1175

- Remove Size_ field
- Validate From/Hash fields in ante handler
- Recompute tx hashes in json-rpc apis to cope with old blocks

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->


<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

yihuang commented on CHANGELOG.md:
```diff
@@ -48,6 +48,10 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
+### Bug Fixes
+
+- (rpc) [#]() Fix eth tx hashes in json-rpc responses
```
```suggestion
- (rpc) [#1176](https://github.com/evmos/ethermint/pull/1176) Fix eth tx hashes in json-rpc responses
```

fedekunze reviewed (changes requested):
Thanks @yihuang, please add tests so that this doesn't break again

yihuang commented on CHANGELOG.md:
```diff
@@ -42,6 +42,8 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 * (deps) [\#1159](https://github.com/evmos/ethermint/pull/1159) Bump Geth version to `v1.10.19`.
 * (deps) [#1167](https://github.com/evmos/ethermint/pull/1167) Upgrade ibc-go to v4.
+* (ante) [#1176](https://github.com/evmos/ethermint/pull/1176) Validate Hash/From in ante handler, recompute eth tx
+  hashes in json-rpc APIs.
```
```suggestion
* (ante) [#1176](https://github.com/evmos/ethermint/pull/1176) Fix invalid tx hashes; Remove `Size_` field and validate `Hash`/`From` fields in ante handler, recompute eth tx
  hashes in JSON-RPC APIs.
```

yihuang commented:
> Thanks @yihuang, please add tests so that this doesn't break again

unit tests added.
And I updated the PR to fix the root cause:
- Removed `Size_` field from `MsgEthereumTx`
- Validate `Hash/From` field in `MsgEthereumTx`

yihuang commented on app/ante/eth.go:
```diff
@@ -429,6 +431,12 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 				return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "invalid message type %T, expected %T", msg, (*evmtypes.MsgEthereumTx)(nil))
 			}
 
+			// Validate Hash/From fields.
```
```suggestion
			// Validate Hash fields.
```

yihuang commented on app/ante/eth.go:
```diff
@@ -68,7 +68,9 @@ func (esvd EthSigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, s
 		}
 
 		// set up the sender to the transaction field if not already
```
```suggestion
		// Validate `From` field
```

yihuang commented on app/ante/eth.go:
```diff
@@ -429,6 +431,12 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 				return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "invalid message type %T, expected %T", msg, (*evmtypes.MsgEthereumTx)(nil))
 			}
 
+			// Validate Hash fields.
```
```suggestion
			// Validate Hash field
```

yihuang commented:
> Thanks @yihuang, please add tests so that this doesn't break again

tests added.

fedekunze commented on app/ante/eth.go:
```diff
@@ -61,13 +61,10 @@ func (esvd EthSigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, s
 			)
 		}
-
-		// set up the sender to the transaction field if not already
```
nit, readd comment

yihuang replied:
done

fedekunze commented on app/ante/eth.go:
```diff
@@ -429,6 +426,20 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 			}
 
+			// Validate Size_ field, should be kept empty
+			if msgEthTx.Size_ != 0 {
```
why not check these fields on the `msg.ValidateBasic` since they are stateless?

yihuang replied:
makes sense, it didn't run `msg.ValidateBasic` at all before, I've just added that.

yihuang replied:
Not sure if not running `msg.ValidateBasic` is expected: https://github.com/cosmos/cosmos-sdk/issues/12607

fedekunze commented on app/ante/eth.go:
```diff
@@ -424,11 +423,20 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 				return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "invalid message type %T, expected %T", msg, (*evmtypes.MsgEthereumTx)(nil))
 			}
 
+			// Validate `From` field
+			if msgEthTx.From != "" {
+				return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid From %s, expect empty string", msgEthTx.From)
+			}
```
I thought this was checked on the message valiadation?

yihuang replied:
message validation will accept non empty `From`, because it's used in places other than ante handler, but in ante handler we want more strict.

fedekunze commented on x/evm/types/msg.go:
```diff
@@ -337,6 +347,7 @@ func (msg *MsgEthereumTx) BuildTx(b client.TxBuilder, evmDenom string) (signing.
 	}
 
 	builder.SetExtensionOptions(option)
+	msg.From = ""
```
nit, comment

yihuang replied:
done

fedekunze commented on x/evm/spec/04_transactions.md:
```diff
@@ -14,7 +14,7 @@ An EVM state transition can be achieved by using the `MsgEthereumTx`. This messa
  // inner transaction data
  Data *types.Any `protobuf:"bytes,1,opt,name=data,proto3" json:"data,omitempty"`
- // encoded storage size of the transaction
+ // DEPRECATED: encoded storage size of the transaction
  Size_ float64 `protobuf:"fixed64,2,opt,name=size,proto3" json:"-"`
```
why not use `reserved`?

yihuang replied:
Because it'll break the tx format, not sure if we want to do a breaking change like that.

yihuang commented on app/ante/eth.go:
```diff
@@ -424,11 +423,20 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 		for _, msg := range protoTx.GetMsgs() {
+			if err := msg.ValidateBasic(); err != nil {
+				return ctx, sdkerrors.Wrap(err, "msg basic validation failed")
+			}
```
done

yihuang commented on app/ante/eth.go:
```diff
@@ -471,6 +479,8 @@ func (vbd EthValidateBasicDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simu
 			return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "for eth tx Signatures should be empty")
 		}
+	} else {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "invalid tx type %T, didn't implement interface protoTxProvider", tx)
```
sounds good, done

yihuang commented on proto/ethermint/evm/v1/tx.proto:
```diff
@@ -25,7 +25,7 @@ message MsgEthereumTx {
   // caches
 
-  // encoded storage size of the transaction
+  // DEPRECATED: encoded storage size of the transaction
   double size = 2 [ (gogoproto.jsontag) = "-" ];
```
That'll breaking current using tx format, worry if it's too big a breaking change.

yihuang commented:
Added a commit to workaround the issue in event parsing, this is needed to handle the legacy blocks even if we have the root cause fix, without require user to resync from scatch.

fedekunze reviewed (approved):
ACK. Let's create an issue to deal with the breaking change

fedekunze commented:
@yihuang some checks are failing

yihuang commented:
> ACK. Let's create an issue to deal with the breaking change

Do you mean removing the `Size_` field?

yihuang commented:
> @yihuang some checks are failing

done, unit tests fixed.
